docs/object-capabilities/derived-authority.md

```diff
@@ -0,0 +1,109 @@
+  fun ref connected(conn: TCPConnection ref) =>
+    _out.print("connected")
+    conn.close()
+
+  fun ref connect_failed(conn: TCPConnection ref) =>
+    _out.print("connect_failed")
+
+actor Connect
+  new create(out: OutStream, auth: AmbientAuth) =>
+    TCPConnection(auth, MyTCPConnectionNotify(out), "example.com", "80")
+
+actor Main
+  new create(env: Env) =>
+    try Connect(env.out, env.root as AmbientAuth) end
+```
+
+The Main actor authorizes the Connect actor by passing the ambient authority token on to it. This token is not forgeable since the `AmbientAuth` constructor is private—only the runtime glue code
+can create the one and only instance.
+
+The Connect actor uses this authority when it creates a TCP connection:
+
+```pony
+TCPConnection(auth, MyTCPConnectionNotify(out), "example.com", "80")
+```
+
+The TCPConnection requires an authority as first parameter, and since the compiler checks that
+the correct type was passed, this guarantees that a TCPConnection can only be created by an
+actor holding the required authorization. If you look into the implementation of the `TCPConnection`
+constructor, you will notice that the authorization is not even used beyond the declaration of the
+parameter.
+
+## Restrict, then delegate your authority
+
+In order to handle our own code and that of others more safely, and also to understand our code better,
+we want to split up the authority, and only grant the particular authority a piece of code actually
+requires.
+
+Imagine we don't trust the `Connect` actor, so we don't want to provide it with more authority
+than needed. For example, there is no point in granting it filesystem access.
+
+The first parameter of the `TCPConnection` constructor has the type
+
+```pony
+type TCPConnectionAuth is (AmbientAuth | NetAuth | TCPAuth | TCPConnectAuth)
+```
+
+This looks like a nice hierarchy of authority, in order of decreasing rights. We can pass
+a reduced authority to our `Connect` actor like this:
+
+```pony
+actor Connect
+  new create(out: OutStream, auth: TCPConnectionAuth) =>
+    TCPConnection(auth, MyTCPConnectionNotify(out), "example.com", "80")
+
+actor Main
+  new create(env: Env) =>
+    try Connect(env.out, TCPConnectAuth(env.root as AmbientAuth)) end
+```
+
+Now we are sure it cannot access the filesystem or listen on a TCP or UDP port. This is possible
+because there is a `TCPConnectAuth` constructor which accepts the AmbientAuth.
+
+You can see this pattern over and over in the standard library.
+
+## Authorization-friendly interface
+
+Consider the above example again, but this time let's treat the `Connect` actor as coming from
+a 3rd party package. If the author of that package had not considered which authority exactly
+was necessary, it would not be possible to restrict the authoritization given to its code.
+A function with an `AmbientAuth` parameter will not accept a lesser auth token.
+
+Therefore it is necessary to design your packages and functions with this in mind.
```

Comment:
Can you logically flip this around? By that I mean rather than placing the reader as using an outside package, place them as writing the outside package. This edit is a preference request as there is nothing wrong with how it is written, but I think switching it around will let you more cleanly and clearly jump off from the previous subsection where we made a change to the **implementation** rather than a change to how the implementation is used.

Reply:
I did that and also expanded a bit. Not sure the introduction of another level of type aliasing is really useful though.